analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Branching.LearnConstraint.cs

```diff
@@ -887,6 +887,22 @@ public void Branching_IsNullOperation_Equals(string expression)
                 .And.ContainSingle(x => x.HasConstraint(ObjectConstraint.NotNull));
         }
 
+        [TestMethod]
+        public void Branching_LearnIsNullSurvivesLambdaCapture()
+        {
+            var validator = SETestContext.CreateCS("""
+                if (s == null)
+                {
+                    return;
+                }
+
+                Func<string> someFunc = () => s.ToString();
+
+                Tag("End", s);
+                """, ", string s").Validator;
```

Comment:
@pavel-mikula-sonarsource This is what I found out so far, but I have no idea how to fix it:

When we enter the block after the return, `s` has the NotNull constraint as expected. `s` is also in the `captured` list of `LiveVariableAnalysisBase`. Therefore it is not included in the `LiveOut` list:
```cs
        /// <summary>
        /// LiveOut variables are alive when exiting block. They are read in any of it's successors.
        /// </summary>
        public IReadOnlyCollection<ISymbol> LiveOut(TBlock block) =>
            blockLiveOut[block].Except(captured).ToImmutableArray();
```

And that is why it gets removed in `ProcessBranchState`:

https://github.com/SonarSource/sonar-dotnet/blob/0d3caf30c09f7c8e99ab783838e186dd72bf572c/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RoslynSymbolicExecution.cs#L165-L167

I don't understand why captured variables are not included in LiveOut and therefore I don't know how to fix it without breaking other stuff. @pavel-mikula-sonarsource, please give me directions on how to fix this.

Reply:
This is tricky. 
LVA tracks `Captured` so we can say "We don't assume anything about them".
The old engine did not track those symbols at all.
I don't think we can properly fix it until we add proper support for lambdas and local functions. As S3900 is not a SonarWay, I would add this reproducer, accept it as FP for this sprint and create a new issue for it (if we don't have similar one).

And in the issue, capture as much information as you've gathered here.

Reply:
New issue created and added to the backlog: #7060